fabric-data-attachment-api-v1/src/main/java/net/fabricmc/fabric/impl/attachment/AttachmentTypeImpl.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.impl.attachment;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
+
+@ApiStatus.Internal
+public class AttachmentTypeImpl<T> implements AttachmentType<T> {
+	private static Map<Identifier, AttachmentType<?>> attachmentsMap = new HashMap<>();
+
+	public static <T> AttachmentType<T> create(Identifier identifier, Class<T> attachmentClass, @Nullable Serializer<T> serializer) {
+		Objects.requireNonNull(identifier, "Identifier may not be null.");
+		Objects.requireNonNull(attachmentClass, "Attachment class may not be null.");
+
+		synchronized (AttachmentTypeImpl.class) {
+			if (attachmentsMap.containsKey(identifier)) {
+				throw new IllegalArgumentException("Duplicate AttachmentType creation for identifier " + identifier);
+			}
+
+			// Use copy-on-write here to ensure fast lock-free queries
+			Map<Identifier, AttachmentType<?>> newMap = new HashMap<>(attachmentsMap.size() + 1);
+			newMap.putAll(attachmentsMap);
+			AttachmentTypeImpl<T> attachment = new AttachmentTypeImpl<>(identifier, attachmentClass, serializer);
+			newMap.put(identifier, attachment);
+			attachmentsMap = newMap;
+			return attachment;
+		}
+	}
+
+	@Nullable
+	public static AttachmentType<?> get(Identifier identifier) {
+		return attachmentsMap.get(identifier);
```

Comment:
This is not thread safe without synchronization

Reply:
Right, I forgot it should be volatile.